### PR TITLE
Workaround issue with Rambler IMAP provider.

### PR DESCRIPTION
**Pull request by disaykin**
Occasionally, for large attachments (~20MB) Rambler returns wrong length of the literal.

See example below (real example attached):

C: 1 UID FETCH 2 BODY.PEEK[2]
S: * 140 FETCH (BODY[2] {100}\r\n
     123456789 UID 2)\r\n
S: 1 OK completed\r\n

As you can see server states '100' as a length of the literal, but provides only 9 bytes.

Mailcore2 awaits exactly '100' bytes and ignores the fact that server has completed his response.

Two issues arise in result:

* Response will not be parsed as it's size < 100 bytes. Mailcore2 will wait for the remaining bytes with no result and block one connection from the pool for 30 seconds (timeout).
* Number of subsequent requests with urgent=NO may be placed into hanged connection queue.

### CODE REVIEW

File: src/data-types/mailstream.c [hunk 1]
```diff
@@ -246,12 +218,13 @@ ssize_t mailstream_read(mailstream * s, void * buf, size_t count)
   }
 
   if (left > s->buffer_max_size) {
-    read_bytes = read_through_buffer(s, cur_buf, left);
+    read_bytes = mailstream_low_read(s->low, cur_buf, left);
```
Comment (dinhviethoa) [inline src/data-types/mailstream.c]: I'm not sure why you're making that change.

File: src/data-types/mailstream_helper.c [hunk 2]
```diff
@@ -62,16 +62,21 @@ char * mailstream_read_line(mailstream * stream, MMAPString * line)
   return mailstream_read_line_append(stream, line);
 }
 
-static char * mailstream_read_len_append(mailstream * stream,
-					 MMAPString * line,
-					 size_t i)
+static char * mailstream_read_len_append_from_buffer(mailstream * stream,
```
Comment (dinhviethoa) [inline src/data-types/mailstream_helper.c]: I'm not sure to understand that change here.

Comment (disaykin) [inline src/data-types/mailstream_helper.c]: The mailstream_read_len_append function always reads from buffer. I try reflect this fact in their name.
I undo this change.

File: src/data-types/mailstream.c (hunk 1 above)
Comment (disaykin) [inline src/data-types/mailstream.c]: This function (read_through_buffer) is always trying to read exactly \count bytes. If the data comes less than needed, the call will end with timeout.  Instead, we may wait for the first piece of data and return it to the caller. If the caller needs some more data, mailstream_read() will be called again.

**Comment by disaykin** [comment]
Next try)

**Comment by disaykin** [comment]
Hi! Can you tell us what do you think about this pull request?

File: src/data-types/mailstream.c (hunk 1 above)
Comment (dinhviethoa) [inline src/data-types/mailstream.c]: The other usage of mailstream_read() is here: mailstream_read_line_append() but it seems that it's going to be alright here too.

**Comment by dinhviethoa** [comment]
I didn't forecast the addition of the parameter to parser functions even though it's a good way to do it.
I'm also think that maybe we should also add in that context the progress parameters, buffer and mailstream.
I'll need to do a major cleanup here :)

**Comment by dinhviethoa** [comment]
You're not looking for "OK completed" yet in this pull request? Are you planning to do that in a separate one?

**Comment by disaykin** [comment]
No, i am looking for "OK completed" in this patch.
But Github hides those changes with "683 additions, 662 deletions not shown because the diff is too large. Please use a local Git client to view these changes."

**Comment by dinhviethoa** [comment]
I downloaded it:
https://github.com/dinhviethoa/libetpan/pull/240.diff

and it still doesn't show it.
Is it in a specific file?

**Comment by disaykin** [comment]
The current version of the parser is very good. I do not think that it requires much rewriting. Maybe, only if required again to add a workaround...

**Comment by dinhviethoa** [comment]
> The current version of the parser is very good. I do not think that it requires much rewriting. Maybe, only if required again to add a workaround...

And the parser_context you're adding will be really helpful with adding more workaround in the future.

**Comment by disaykin** [comment]
The change is in this hunk:
@@ -6074,7 +6074,28 @@ static int mailimap_literal_parse_progress(mailstream * fd, MMAPString * buffer,
of src/low-level/imap/mailimap_parser.c.

Perhaps I misunderstood what you can not find

**Comment by dinhviethoa** [comment]
Got it:

```
+      if (needed > 0 && mailimap_parser_context_is_rambler_workaround_enabled(parser_ctx)) {
+        /* workaround issue with Rambler IMAP server */
+        char const *literal_end = literal_p;
+        char tag_response_end[] = " OK completed\r\n";
+        size_t tag_response_end_len = sizeof(tag_response_end) - 1;
+        if (literal->len > tag_response_end_len && !memcmp(literal_end - tag_response_end_len, tag_response_end, tag_response_end_len)) {
```

`literal->len >= tag_response_end_len` would work?
use `memcpy() == 0` instead.

```
+          literal_end -= tag_response_end_len;
+          while (literal_end > literal->str && is_digit(*(literal_end - 1))) literal_end--;
```

Could you add comments about what you're doing here (and below)?

```
+          if (literal_end > literal->str + 3) literal_end -= 3;
+          if (!memcmp(literal_end, ")\r\n", 3)) {
+            while (literal_end > literal->str && is_digit(*(literal_end - 1))) literal_end--;
+            if (literal_end > literal->str + 5) literal_end -= 5;
+            if (!memcmp(literal_end, " UID ", 5)) {
+              /* got invalid server response, no more data will received */
+              res = MAILIMAP_ERROR_STREAM;
```

How about `MAILIMAP_ERROR_PARSE` instead?

```
+              goto free_literal;
+            }
+          }
+        }
+      }
+
```

**Comment by disaykin** [comment]
> How about MAILIMAP_ERROR_PARSE instead?
This error type used by parser for utility purposes. The parser will try apply next pattern after receiving MAILIMAP_ERROR_PARSE.

**Comment by disaykin** [comment]
Done.

**Comment by dinhviethoa** [comment]
> This error type is used by parser for utility purposes. The parser will try apply next pattern after receiving MAILIMAP_ERROR_PARSE.

My thought is that it's going to go through couple of parsing failure and will spit out a parsing failure at the end. The main parsing function will then return MAILIMAP_ERROR_PARSE.

**Comment by disaykin** [comment]
This is not parsing error. This is server error. After we are detecting this server error, all of we can do, it is close the connection.

**Comment by dinhviethoa** [comment]
Parse_error for Liberian generally means that the server did a bad job at producing a good response.
That's your description of server error.
At this point, of course, the best course for the user of the API is to close the connection.

**Comment by disaykin** [comment]
Done.

**Comment by dinhviethoa** [comment]
Merging now :)
Thanks!

**Comment by disaykin** [comment]
Thanks!

**Comment by dinhviethoa** [comment]
@disaykin could you make it happen automatically on mailcore2?
It would be awesome!

See here a way to detect the server:
https://github.com/MailCore/mailcore2/commit/8d1f80ed93d5c23a74705e1627b71cef103e067a
